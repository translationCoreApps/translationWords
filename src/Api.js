import {ToolApi} from 'tc-tool';
import path from 'path-extra';

export default class Api extends ToolApi {

  constructor() {
    super();
    this.getAlignmentMemory = this.getAlignmentMemory.bind(this);
    this.getInvalidChecks = this.getInvalidChecks.bind(this);
    this.getProgress = this.getProgress.bind(this);
    this._loadBookSelections = this._loadBookSelections.bind(this);
    this._loadVerseSelections = this._loadVerseSelections.bind(this);
    this._loadCheckData = this._loadCheckData.bind(this);
  }

  /**
   * Lifecycle method
   */
  toolWillConnect() {
    // TODO: implement
  }

  /**
   * Lifecycle method
   * @param nextState
   * @param prevState
   */
  stateChangeThrottled(nextState, prevState) {
    // TODO: implement
  }

  /**
   * Lifecycle method
   * @param state
   * @param props
   */
  mapStateToProps(state, props) {
    // TODO: implement
  }

  /**
   * Lifecycle method
   * @param dispatch
   */
  mapDispatchToProps(dispatch) {
    // TODO: implement
    return {};
  }

  /**
   * Lifecycle method
   */
  toolWillDisconnect() {
    // TODO: implement
  }

  toolWillReceiveProps(nextProps) {
    // TODO: implement
  }

  /**
   * Loads the most recent check data for the context id
   * @param {string} check - the check to look up e.g. "invalidated"
   * @param {object} contextId - the context id for the check to look up
   * @returns {object|null} - the check data record or null if one cannot be found
   * @private
   */
  _loadCheckData(check, contextId) {
    const {tc: {project}} = this.props;
    const {reference: {bookId, chapter, verse}, groupId, quote, occurrence} = contextId;
    const loadPath = path.join('checkData', check, bookId, `${chapter}`,
      `${verse}`);

    if (project.dataPathExistsSync(loadPath)) {
      // sort and filter check records
      const files = project.readDataDirSync(loadPath).filter(file => {
        return path.extname(file) === '.json';
      });
      let sortedRecords = files.sort().reverse();

      // load check data
      for (let i = 0, len = sortedRecords.length; i < len; i++) {
        const record = sortedRecords[i];
        const recordPath = path.join(loadPath, record);
        try {
          const recordData = JSON.parse(project.readDataFileSync(recordPath));

          // return first match
          if (recordData.contextId.groupId === groupId &&
            recordData.contextId.quote === quote &&
            recordData.contextId.occurrence === occurrence) {
            return recordData;
          }
        } catch (e) {
          console.warn(`Failed to load check record from ${recordPath}`, e);
        }
      }
    }

    return null;
  }

  /**
   * Returns the total number of invalided checks
   * TODO: move category selection management into the tool so we don't need this param
   * @param {string[]} selectedCategories - an array of categories to include in the calculation.
   * @returns {number} - the number of invalid checks
   */
  getInvalidChecks(selectedCategories) {
    const {tc: {project}, tool: {name}} = this.props;
    let invalidChecks = 0;

    for (const category of selectedCategories) {
      const groups = project.getCategoryGroupIds(name, category);
      for (const group of groups) {
        const data = project.getGroupData(name, group);
        if (data && data.constructor === Array) {
          for (const check of data) {
            const checkData = this._loadCheckData('invalidated',
              check.contextId);
            if (checkData && checkData.invalidated === true) {
              invalidChecks++;
            }
          }
        } else {
          console.warn(`Invalid group data found for "${group}"`);
        }
      }
    }

    return invalidChecks;
  }

  /**
   * Returns the percent progress of completion for the project.
   * TODO: move category selection management into the tool so we don't need this param
   * @param {string[]} selectedCategories -  an array of categories to include in the calculation.
   * @returns {number} - a value between 0 and 1
   */
  getProgress(selectedCategories) {
    const {tc: {project}, tool: {name}} = this.props;
    let totalChecks = 0;
    let completedChecks = 0;

    for (const category of selectedCategories) {
      const groups = project.getCategoryGroupIds(name, category);
      for (const group of groups) {
        const data = project.getGroupData(name, group);
        if (data && data.constructor === Array) {
          for (const check of data) {
            totalChecks++;
            completedChecks += check.selections ? 1 : 0;
          }
        } else {
          console.warn(`Invalid group data found for "${group}"`);
        }
      }
    }

    if (totalChecks === 0) {
      return 0;
    } else {
      return completedChecks / totalChecks;
    }
  }

  /**
   * Returns the alignment memory generated from selections made in tW.
   * @return {{sourceText : string, targetText : string}[]}
   */
  getAlignmentMemory() {
    // TODO: perform initial selection loading when the tool connects.
    // This must wait until after all selection logic is migrated to tW.
    const selections = this._loadBookSelections(this.props);
    const alignmentMemory = [];

    // format selections as alignment memory
    for (const chapter of Object.keys(selections)) {
      for (const verse of Object.keys(selections[chapter])) {
        for (const selection of selections[chapter][verse]) {
          if (selection.selections.length === 0) continue;
          const sourceText = selection.contextId.quote;
          const targetText = selection.selections.map(s => s.text).join(' ');
          alignmentMemory.push({
            sourceText,
            targetText
          });
        }
      }
    }

    return alignmentMemory;
  }

  /**
   * Loads the selection data for the entire book
   * @param props
   * @private
   */
  _loadBookSelections(props) {
    const {
      tc: {
        targetBook
      }
    } = props;

    const selections = {};
    for (const chapter of Object.keys(targetBook)) {
      for (const verse of Object.keys(targetBook[chapter])) {
        const verseSelections = this._loadVerseSelections(chapter, verse,
          props);
        if (verseSelections.length > 0) {
          if (!selections[chapter]) {
            selections[chapter] = {};
          }
          selections[chapter][verse] = verseSelections;
        }
      }
    }
    // TODO: store selection data in tool reducer.
    return selections;
  }

  /**
   * Loads the selection data for a verse
   * @param {string} chapter
   * @param {string} verse
   * @param props
   * @return {Array}
   * @private
   */
  _loadVerseSelections(chapter, verse, props) {
    const {
      tc: {
        contextId: {reference: {bookId}},
        projectDataPathExistsSync,
        readProjectDataSync,
        readProjectDataDirSync
      }
    } = props;

    const verseDir = path.join('checkData/selections/', bookId, chapter, verse);
    const selections = [];
    const foundSelections = [];
    if (projectDataPathExistsSync(verseDir)) {

      let files = readProjectDataDirSync(verseDir);
      files = files.filter(f => path.extname(f) === '.json');
      files = files.sort().reverse();
      for (let i = 0; i < files.length; i++) {
        let filePath = path.join(verseDir, files[i]);

        let data;
        try {
          data = JSON.parse(readProjectDataSync(filePath));
        } catch (err) {
          console.error(`failed to read selection data from ${filePath}`, err);
          continue;
        }

        if (data && data.contextId) {
          const id = `${data.contextId.groupId}:${data.contextId.quote}`;
          if (foundSelections.indexOf(id) === -1) {
            foundSelections.push(id);
            selections.push(data);
          }
        }
      }
    }
    return selections;
  }

}
