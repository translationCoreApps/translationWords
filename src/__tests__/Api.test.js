/* eslint-env jest */
import Api from '../Api';

describe('load check data', () => {
  it('returns older data in favor of context match', () => {
    const api = new Api();
    const contextId = {
      reference: {bookId: 'tit', chapter: 1, verse: 1},
      groupId: 'figs_metaphor',
      quote: 'that he put before them',
      occurrence: 1
    };
    const project = {
      dataPathExistsSync: jest.fn(),
      readDataDirSync: jest.fn(),
      readDataFileSync: jest.fn()
    };
    api.props = {
      tc: {
        project
      }
    };

    project.dataPathExistsSync.mockReturnValueOnce(true);
    project.readDataDirSync.mockReturnValueOnce(
      ['2018-12-18T21_28_18.837Z.json', '2019-01-10T03_59_47.588Z.json']);
    project.readDataFileSync.mockReturnValueOnce(
      JSON.stringify(
        {contextId: {groupId: 'hmmm', quote: 'hello', occurrence: 2}}));
    project.readDataFileSync.mockReturnValueOnce(
      JSON.stringify(
        {
          contextId: {
            groupId: 'figs_metaphor',
            quote: 'that he put before them',
            occurrence: 1
          }
        }));

    const data = api._loadCheckData('invalidated', contextId);
    expect(data).toMatchSnapshot();
    expect(project.readDataFileSync.mock.calls[0]).
      toEqual(
        ['checkData/invalidated/tit/1/1/2019-01-10T03_59_47.588Z.json']);
  });

  it('returns latest data in favor of context match', () => {
    const api = new Api();
    const contextId = {
      reference: {bookId: 'tit', chapter: 1, verse: 1},
      groupId: 'figs_metaphor',
      quote: 'that he put before them',
      occurrence: 1
    };
    const project = {
      dataPathExistsSync: jest.fn(),
      readDataDirSync: jest.fn(),
      readDataFileSync: jest.fn()
    };
    api.props = {
      tc: {
        project
      }
    };

    project.dataPathExistsSync.mockReturnValueOnce(true);
    project.readDataDirSync.mockReturnValueOnce(
      ['2018-12-18T21_28_18.837Z.json', '2019-01-10T03_59_47.588Z.json']);
    project.readDataFileSync.mockReturnValueOnce(
      JSON.stringify(
        {
          contextId: {
            groupId: 'figs_metaphor',
            quote: 'that he put before them',
            occurrence: 1
          }
        }));
    project.readDataFileSync.mockReturnValueOnce(
      JSON.stringify(
        {contextId: {groupId: 'hmmm', quote: 'hello', occurrence: 2}}));
    const data = api._loadCheckData('invalidated', contextId);
    expect(data).toMatchSnapshot();
    expect(project.readDataFileSync.mock.calls[0]).
      toEqual(
        ['checkData/invalidated/tit/1/1/2019-01-10T03_59_47.588Z.json']);
  });
});

describe('alignment memory', () => {
  it('returns the alignment memory', () => {
    const api = new Api();
    api._loadBookSelections = jest.fn(() => {
      return {
        '1': {
          '1': [
            {
              contextId: {
                quote: 'hello'
              },
              selections: [
                {
                  text: 'world'
                }
              ]
            }
          ]
        }
      };
    });
    const memory = api.getAlignmentMemory();
    expect(memory).toEqual([
      {
        sourceText: 'hello',
        targetText: 'world'
      }
    ]);
  });
});

describe('selection data', () => {

  it('loads selection data from the disk', () => {
    // Replicates the file reader
    function* dataGenerator() {
      const data2 = JSON.stringify({
        contextId: {
          groupId: 'word',
          quote: 'hi'
        },
        selections: [
          {
            text: 'world'
          }
        ]
      });

      const data1 = JSON.stringify({
        contextId: {
          groupId: 'word',
          quote: 'hello'
        },
        selections: [
          {
            text: 'world'
          }
        ]
      });
      // TRICKY: files will be sorted in ascending order
      yield data2;
      yield data1;
      yield data1;
    }

    const api = new Api();
    const generator = dataGenerator();
    const props = {
      tc: {
        contextId: {reference: {bookId: 'book'}},
        projectDataPathExistsSync: jest.fn(() => true),
        readProjectDataSync: jest.fn(() => generator.next().value),
        readProjectDataDirSync: jest.fn(() => {
          return ['1.json', '1_dup.json', '2.json'];
        }),
        targetBook: {
          '1': {
            '1': {}
          }
        }
      }
    };
    const selections = api._loadBookSelections(props);
    expect(selections).toEqual({
      '1': {
        '1': [
          {
            contextId: {
              groupId: 'word',
              quote: 'hi'
            },
            selections: [
              {
                text: 'world'
              }
            ]
          },
          {
            contextId: {
              groupId: 'word',
              quote: 'hello'
            },
            selections: [
              {
                text: 'world'
              }
            ]
          }
        ]
      }
    });
  });
});
