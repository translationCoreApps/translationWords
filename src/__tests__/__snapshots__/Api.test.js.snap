// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load check data returns latest data in favor of context match 1`] = `
Object {
  "contextId": Object {
    "groupId": "figs_metaphor",
    "occurrence": 1,
    "quote": "that he put before them",
  },
}
`;

exports[`load check data returns older data in favor of context match 1`] = `
Object {
  "contextId": Object {
    "groupId": "figs_metaphor",
    "occurrence": 1,
    "quote": "that he put before them",
  },
}
`;
